### PR TITLE
cucumber-cpp 0.3 (new formula)

### PR DESCRIPTION
**Pull request by MartinDelille**
### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew/pulls) for the same update/change?

### New Formulae Submissions:

- [x] Does your submission pass
`brew audit --strict --online <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [x] Have you built your formula locally prior to submission with `brew install <formula>`?

As you can see I didn't check all because `brew audit --strict --online cucumber-cpp` doesn't succeed and give me the following output:

```
$ brew audit --strict --online cucumber-cpp
==> brew style cucumber-cpp

1 file inspected, no offenses detected

==> audit problems
cucumber-cpp:
 * Don't use ruby as a dependency. We allow non-Homebrew ruby installations.

Error: 1 problem in 1 formula
```

I tried to remove `depends_on "ruby"` but then I have the following error when trying to install:

```
==> gem install bundler
ERROR:  While executing gem ... (Gem::FilePermissionError)
    You don't have write permissions for the /Library/Ruby/Gems/2.0.0 directory.
```

I submit it hopping that someone could help solving this issue.

### CODE REVIEW

File: Library/Formula/cucumber-cpp.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  homepage "https://cucumber.io"
+  url "https://github.com/cucumber/cucumber-cpp/archive/1e633c6b59c75b959b480dcc4e40123e7fc59032.tar.gz"
+  version "0.3"
+  sha256 "1e211db8b8c230f291c8774d1c2e2bbbcf264d92016fd6081557425e43569c19"
+
+  depends_on "cmake" => :build
+  depends_on "ruby"
+
+  def install
+    system "gem", "install", "bundler"
```
Comment (DomT4) [inline Library/Formula/cucumber-cpp.rb]: Why is this needed? We can't do it this way because it'll try and do a system-wide install, which violates our sandboxing.

File: Library/Formula/cucumber-cpp.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    system "cmake", "-DCUKE_DISABLE_GTEST=on",
+                    "-DCUKE_DISABLE_CPPSPEC=on",
+                    "."
+    system "cmake", "--build", "."
+    include.install "include/cucumber-cpp"
+    lib.install Dir["src/*.a"]
+  end
+
+  test do
+    system "false"
```
Comment (DomT4) [inline Library/Formula/cucumber-cpp.rb]: Can this test be modified to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

File: Library/Formula/cucumber-cpp.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
+class CucumberCpp < Formula
+  desc "Support for writing Cucumber step definitions in C++"
+  homepage "https://cucumber.io"
+  url "https://github.com/cucumber/cucumber-cpp/archive/1e633c6b59c75b959b480dcc4e40123e7fc59032.tar.gz"
```
Comment (DomT4) [inline Library/Formula/cucumber-cpp.rb]: We'd want to use an actual stable, tagged release here.

File: Library/Formula/cucumber-cpp.rb (hunk 1 above)
Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: In fact this would be needed just for (currently empty) testing purpose. Once a program is compiled with this library, it'll start listening on the port 3902, we need to execute the `cucumber` gem against it.

File: Library/Formula/cucumber-cpp.rb (hunk 2 above)
Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: I will fix this but I need the `cucumber` gem installed for the test, how can I do that?

File: Library/Formula/cucumber-cpp.rb (hunk 3 above)
Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: Ok I will use the last release.

File: Library/Formula/cucumber-cpp.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  version "0.3"
+  sha256 "1e211db8b8c230f291c8774d1c2e2bbbcf264d92016fd6081557425e43569c19"
+
+  depends_on "cmake" => :build
+  depends_on "ruby"
```
Comment (ilovezfs) [inline Library/Formula/cucumber-cpp.rb]: why is ruby an unconditional dependency

Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: As I said in a previous comment:

> In fact this would be needed just for (currently empty) testing purpose. Once a program is compiled with this library, it'll start listening on the port 3902, we need to execute the cucumber gem against it.

File: Library/Formula/cucumber-cpp.rb (hunk 1 above)
Comment (ilovezfs) [inline Library/Formula/cucumber-cpp.rb]: You cannot use the system ruby?

Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: No because I need to install the `cucumber` gem and it leads to a permission error. How can I have a gem dependency? It is described there but without much details: https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/Formula-Cookbook.md#specifying-gems-python-modules-go-projects-etc-as-dependencies

File: Library/Formula/cucumber-cpp.rb (hunk 2 above)
Comment (ilovezfs) [inline Library/Formula/cucumber-cpp.rb]:
```
brew install --interactive --debug cucumber-cpp
/usr/bin/irb
puts Gem.path
```

So one way is to put the resource in the make-shift home directory, which should work.

For <= Mountain Lion you'd want to skip the test, or make homebrew ruby a hard dependency for <= Mountain Lion, since cucumber requires ruby 1.9.3

**Comment by mikemcquaid** [comment]
We need to work on a way to not call `gem install` or require a non-system Ruby for this.

File: Library/Formula/cucumber-cpp.rb (hunk 2 above)
Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: How?

Comment (mikemcquaid) [inline Library/Formula/cucumber-cpp.rb]: We still haven't figured out how to adequately use `bundler` in Homebrew formulae in a way that ensures the results are vendored only into the `prefix`.

Comment (ilovezfs) [inline Library/Formula/cucumber-cpp.rb]: This seems to work:
```
diff --git a/Library/Formula/cucumber-cpp.rb b/Library/Formula/cucumber-cpp.rb
index c59677d..2b8d00a 100644
--- a/Library/Formula/cucumber-cpp.rb
+++ b/Library/Formula/cucumber-cpp.rb
@@ -6,13 +6,18 @@ class CucumberCpp < Formula
   sha256 "1e211db8b8c230f291c8774d1c2e2bbbcf264d92016fd6081557425e43569c19"
 
   depends_on "cmake" => :build
-  depends_on "ruby"
+  depends_on "boost"
+  #depends_on "ruby"
 
   def install
+    home = ENV["HOME"]
+    ENV["GEM_HOME"] = home
+    ENV["BUNDLE_PATH"] = home
+    ENV.prepend_path "PATH", "#{home}/gems/bundler-1.11.2/exe"
+    ENV.prepend_path "PATH", "#{home}/gems/bundler-1.11.2/bin"
     system "gem", "install", "bundler"
-    system "cmake", "-DCUKE_DISABLE_GTEST=on",
-                    "-DCUKE_DISABLE_CPPSPEC=on",
-                    "."
+    system "cmake", *std_cmake_args, "-DCUKE_DISABLE_GTEST=on",
+                    "-DCUKE_DISABLE_CPPSPEC=on"
     system "cmake", "--build", "."
     include.install "include/cucumber-cpp"
     lib.install Dir["src/*.a"]
```

However, the build eventually dies.

```
[ 39%] Building CXX object src/CMakeFiles/cucumber-cpp.dir/connectors/wire/WireProtocolCommands.cpp.o
cd /tmp/cucumber-cpp20160222-7926-1id94rm/cucumber-cpp-1e633c6b59c75b959b480dcc4e40123e7fc59032/src && /usr/local/Library/ENV/4.3/clang++    -I/usr/local/include -I/tmp/cucumber-cpp20160222-7926-1id94rm/cucumber-cpp-1e633c6b59c75b959b480dcc4e40123e7fc59032/include  -DBOOST_ALL_DYN_LINK -DNDEBUG   -o CMakeFiles/cucumber-cpp.dir/connectors/wire/WireProtocolCommands.cpp.o -c /tmp/cucumber-cpp20160222-7926-1id94rm/cucumber-cpp-1e633c6b59c75b959b480dcc4e40123e7fc59032/src/connectors/wire/WireProtocolCommands.cpp
[ 42%] Building CXX object src/CMakeFiles/cucumber-cpp.dir/drivers/BoostDriver.cpp.o
cd /tmp/cucumber-cpp20160222-7926-1id94rm/cucumber-cpp-1e633c6b59c75b959b480dcc4e40123e7fc59032/src && /usr/local/Library/ENV/4.3/clang++    -I/usr/local/include -I/tmp/cucumber-cpp20160222-7926-1id94rm/cucumber-cpp-1e633c6b59c75b959b480dcc4e40123e7fc59032/include  -DBOOST_ALL_DYN_LINK -DNDEBUG   -o CMakeFiles/cucumber-cpp.dir/drivers/BoostDriver.cpp.o -c /tmp/cucumber-cpp20160222-7926-1id94rm/cucumber-cpp-1e633c6b59c75b959b480dcc4e40123e7fc59032/src/drivers/BoostDriver.cpp
/tmp/cucumber-cpp20160222-7926-1id94rm/cucumber-cpp-1e633c6b59c75b959b480dcc4e40123e7fc59032/src/drivers/BoostDriver.cpp:83:21: error: no member named 'is_initialized' in namespace 'boost::unit_test::framework'
    if (!framework::is_initialized()) {
         ~~~~~~~~~~~^
/tmp/cucumber-cpp20160222-7926-1id94rm/cucumber-cpp-1e633c6b59c75b959b480dcc4e40123e7fc59032/src/drivers/BoostDriver.cpp:87:30: error: allocating an object of abstract class type 'cucumber::internal::CukeBoostLogInterceptor'
        logInterceptor = new CukeBoostLogInterceptor;
                             ^
/usr/local/include/boost/test/unit_test_log_formatter.hpp:164:25: note: unimplemented pure virtual method 'log_exception_start' in 'CukeBoostLogInterceptor'
    virtual void        log_exception_start( std::ostream& os, log_checkpoint_data const& lcd, execution_exception const& ex ) = 0;
                        ^
/usr/local/include/boost/test/unit_test_log_formatter.hpp:171:25: note: unimplemented pure virtual method 'log_exception_finish' in 'CukeBoostLogInterceptor'
    virtual void        log_exception_finish( std::ostream& os ) = 0;
                        ^
/usr/local/include/boost/test/unit_test_log_formatter.hpp:222:25: note: unimplemented pure virtual method 'entry_context_start' in 'CukeBoostLogInterceptor'
    virtual void        entry_context_start( std::ostream& os, log_level l ) = 0;
                        ^
/usr/local/include/boost/test/unit_test_log_formatter.hpp:230:25: note: unimplemented pure virtual method 'log_entry_context' in 'CukeBoostLogInterceptor'
    virtual void        log_entry_context( std::ostream& os, const_string value ) = 0;
                        ^
/usr/local/include/boost/test/unit_test_log_formatter.hpp:236:25: note: unimplemented pure virtual method 'entry_context_finish' in 'CukeBoostLogInterceptor'
    virtual void        entry_context_finish( std::ostream& os ) = 0;
                        ^
2 errors generated.
make[2]: *** [src/CMakeFiles/cucumber-cpp.dir/drivers/BoostDriver.cpp.o] Error 1
make[1]: *** [src/CMakeFiles/cucumber-cpp.dir/all] Error 2
make: *** [all] Error 2

```

Comment (mikemcquaid) [inline Library/Formula/cucumber-cpp.rb]: @ilovezfs Nice work. Presumably we'd also need to have `bundler` as a `resource`?

Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: I will try that but I think I'll need only the *cucumber* gem.

Comment (ilovezfs) [inline Library/Formula/cucumber-cpp.rb]: @mikemcquaid It seems willing to do it on its own just fine. See below.
@MartinDelille Right. It handles that:

```
==> gem install bundler
Successfully installed bundler-1.11.2
Parsing documentation for bundler-1.11.2
Installing ri documentation for bundler-1.11.2
1 gem installed
==> /private/tmp/cucumber-cpp20160222-18652-ddfdea/cucumber-cpp-1e633c6b59c75b959b480dcc4e40123e7fc59032/.brew_home/bin/bundle
Fetching gem metadata from https://rubygems.org/..........
Fetching version metadata from https://rubygems.org/..
Resolving dependencies...
Rubygems 2.0.14 is not threadsafe, so your gems will be installed one at a time. Upgrade to Rubygems 2.1.0 or higher to enable parallel gem installation.
Installing ffi 1.9.10 with native extensions
Installing builder 3.2.2
Installing diff-lcs 1.2.5
Installing multi_json 1.11.2
Installing multi_test 0.1.2
Installing rspec-core 2.14.8
Installing rspec-mocks 2.14.6
Using bundler 1.11.2
Installing childprocess 0.5.9
Installing rspec-expectations 2.14.5
Installing gherkin 2.12.2 with native extensions
Installing rspec 2.14.1
Installing cucumber 1.3.20
Installing aruba 0.6.1
Bundle complete! 3 Gemfile dependencies, 14 gems now installed.
Bundled gems are installed into ./.brew_home.
==> /private/tmp/cucumber-cpp20160222-18652-ddfdea/cucumber-cpp-1e633c6b59c75b959b480dcc4e40123e7fc59032/.brew_home/bin/bundle show
Gems included by the bundle:
  * aruba (0.6.1)
  * builder (3.2.2)
  * bundler (1.11.2)
  * childprocess (0.5.9)
  * cucumber (1.3.20)
  * diff-lcs (1.2.5)
  * ffi (1.9.10)
  * gherkin (2.12.2)
  * multi_json (1.11.2)
  * multi_test (0.1.2)
  * rspec (2.14.1)
  * rspec-core (2.14.8)
  * rspec-expectations (2.14.5)
  * rspec-mocks (2.14.6)
```

Comment (ilovezfs) [inline Library/Formula/cucumber-cpp.rb]: That's with this diff only:
```
diff --git a/Library/Formula/cucumber-cpp.rb b/Library/Formula/cucumber-cpp.rb
index c59677d..c722370 100644
--- a/Library/Formula/cucumber-cpp.rb
+++ b/Library/Formula/cucumber-cpp.rb
@@ -6,13 +6,18 @@ class CucumberCpp < Formula
   sha256 "1e211db8b8c230f291c8774d1c2e2bbbcf264d92016fd6081557425e43569c19"
 
   depends_on "cmake" => :build
-  depends_on "ruby"
+  depends_on "boost"
 
   def install
+    home = ENV["HOME"]
+    ENV["GEM_HOME"] = home
+    ENV["BUNDLE_PATH"] = home
+    ENV.prepend_path "PATH", "#{home}/bin"
     system "gem", "install", "bundler"
-    system "cmake", "-DCUKE_DISABLE_GTEST=on",
-                    "-DCUKE_DISABLE_CPPSPEC=on",
-                    "."
+    system "#{home}/bin/bundle"
+    system "#{home}/bin/bundle", "show"
+raise "debug"
+    system "cmake", "-DCUKE_DISABLE_GTEST=on", "-DCUKE_DISABLE_CPPSPEC=on", "."
     system "cmake", "--build", "."
     include.install "include/cucumber-cpp"
     lib.install Dir["src/*.a"]

```

**Comment by DomT4** [comment]
How's this going?

**Comment by MartinDelille** [comment]
I didn't had much time to work on this. I'll try to fix it ASAP!

File: Library/Formula/cucumber-cpp.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      GIVEN("^A given statement$") {
+      }
+      WHEN("^A when statement$") {
+      }
+      THEN("^A then statement$") {
+      }
+    EOS
+    system ENV.cxx, "test.cpp", "-L#{lib}", "-lcucumber-cpp", "-o", "test",
+      "-lboost_regex", "-lboost_system"
+#    system "./test&;"
```
Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: How can I launch a process in background like we do with ampersand in the terminal?

File: Library/Formula/cucumber-cpp.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    system "cmake", "--build", "."
+    include.install "include/cucumber-cpp"
+    lib.install Dir["src/*.a"]
+  end
+
+  test do
+    home = ENV["HOME"]
+    ENV["GEM_HOME"] = home
+    ENV["BUNDLE_PATH"] = home
+    ENV.prepend_path "PATH", "#{home}/bin"
```
Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: This `prepend_path` instruction doesn't work.

File: Library/Formula/cucumber-cpp.rb (hunk 5 above)
Comment (DomT4) [inline Library/Formula/cucumber-cpp.rb]: Could fork the process, probably? [Example](https://github.com/Homebrew/homebrew/blob/883c18f4c5941b6b3f5d9d3e6dbab0aa3cae6f17/Library/Formula/serveit.rb#L17-L24)

File: Library/Formula/cucumber-cpp.rb (hunk 6 above)
Comment (DomT4) [inline Library/Formula/cucumber-cpp.rb]: The `home` ENV is always `buildpath`, so probably more obvious to do:
```ruby
ENV["GEM_HOME"] = buildpath
ENV["BUNDLE_PATH"] = buildpath
ENV.prepend_path "PATH", buildpath/"bin"
```
What do you mean by doesn't work, in more detail?

Comment (DomT4) [inline Library/Formula/cucumber-cpp.rb]: Oops, `testpath` here rather than `buildpath`. Apologies. Forgot what block we were in :smile:.

Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: I update my code with your advice but the error is still there: when starting `cucumber` (with `system "cucumber"`) I have an execution failure:

```
==> cucumber
Failed to execute: cucumber
Error: cucumber-cpp: failed
Failed executing: cucumber
/usr/local/Library/Homebrew/formula.rb:1477:in `block in system'
/usr/local/Library/Homebrew/formula.rb:1414:in `open'
/usr/local/Library/Homebrew/formula.rb:1414:in `system'
/usr/local/Library/Formula/cucumber-cpp.rb:52:in `block in <class:CucumberCpp>'
/usr/local/Library/Homebrew/formula.rb:1323:in `block in run_test'
/usr/local/Library/Homebrew/extend/fileutils.rb:35:in `mktemp'
/usr/local/Library/Homebrew/formula.rb:1319:in `run_test'
/usr/local/Library/Homebrew/test.rb:28:in `block in <main>'
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/lib/ruby/2.0.0/timeout.rb:66:in `timeout'
/usr/local/Library/Homebrew/test.rb:27:in `<main>'
```

I manage to make it work by using the full path anyway (system testpath/"cucumber")so it is not a big deal.

File: Library/Formula/cucumber-cpp.rb (hunk 5 above)
Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: :+1:

File: Library/Formula/cucumber-cpp.rb [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  sha256 "1c0f9949627e7528017bf00cbe49693ba9cbc3e11087f70aa33b21df93f341d6"
+
+  depends_on "cmake" => :build
+  depends_on "boost"
+
+  def install
+    system "cmake", "-DCUKE_DISABLE_GTEST=on",
+                    "-DCUKE_DISABLE_CPPSPEC=on",
+                    "-DCUKE_DISABLE_FUNCTIONAL=on",
+                    "-DCUKE_DISABLE_BOOST_TEST=on",
```
Comment (dunn) [inline Library/Formula/cucumber-cpp.rb]: These should be appended to `std_cmake_args`.

**Comment by MartinDelille** [comment]
All check pass! Is there any objection? Otherwise I'll squash the commit.

File: Library/Formula/cucumber-cpp.rb (hunk 7 above)
Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: Done

File: Library/Formula/cucumber-cpp.rb [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      WHEN("^A when statement$") {
+      }
+      THEN("^A then statement$") {
+      }
+    EOS
+    system ENV.cxx, "test.cpp", "-L#{lib}", "-lcucumber-cpp", "-o", "test",
+      "-lboost_regex", "-lboost_system"
+    begin
+      pid = fork { exec "./test" }
+      assert_match %r{^Feature: Test
```
Comment (mikemcquaid) [inline Library/Formula/cucumber-cpp.rb]: Use `EOS.undent` here to mean that you can improve the indentation used.

Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: Ok but the indentation is fine. Shall I just remove `.undent`?

Comment (mikemcquaid) [inline Library/Formula/cucumber-cpp.rb]: Sorry, I don't understand the question?

Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: I don't see how to improve the indentation here. What do you want me to do?

Comment (dunn) [inline Library/Formula/cucumber-cpp.rb]: This is what we mean:
```ruby
      expected = <<-EOS.undent
        Feature: Test

          Scenario: Just for test   # features\/test.feature:2
            Given A given statement # test.cpp:2
            When A when statement   # test.cpp:4
            Then A then statement   # test.cpp:6

        1 scenario \(1 passed\)
        3 steps \(3 passed\)
      EOS
      assert_equal expected, shell_output(testpath/"bin/cucumber")
```

Comment (MartinDelille) [inline Library/Formula/cucumber-cpp.rb]: The only problem is that there is a last line written by cucumber telling the execution time that might not be constant. That's why I wanted to use a regex.

Comment (dunn) [inline Library/Formula/cucumber-cpp.rb]: Oh, in that case `assert_match` should work with a string.

**Comment by MartinDelille** [comment]
Everything seems to be fine now! What's up?
I learned a lot about *brew* formula (and other stuff) with this one! I would like to thank you for the time passing helping me sorting this out :-)

**Comment by mikemcquaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Comment by MartinDelille** [comment]
:+1:
